src/shape.rs

```diff
@@ -348,3 +376,356 @@ mod test
+    #[inline]
+    pub fn bottom(&self) -> T
+    {
+        self.bottom_right.y
+    }
+
+    /// Returns a `Rectangle` representing the rectangle that encloses this
+    /// rounded rectangle.
+    #[inline]
+    pub fn as_rectangle(&self) -> Rectangle<T>
+    {
+        Rectangle::new(self.top_left, self.bottom_right)
+    }
+}
+
+impl<T: std::ops::Sub<Output = T> + Copy> RoundedRectangle<T>
+{
+    /// Returns the width of the rounded rectangle.
+    #[inline]
+    pub fn width(&self) -> T
+    {
+        self.bottom_right.x - self.top_left.x
+    }
+
+    /// Returns the height of the rounded rectangle.
+    #[inline]
+    pub fn height(&self) -> T
+    {
+        self.bottom_right.y - self.top_left.y
+    }
+
+    /// Returns a `Vector2` containing the width and height of the rounded
+    /// rectangle.
+    #[inline]
+    pub fn size(&self) -> Vector2<T>
+    {
+        Vector2::new(self.width(), self.height())
+    }
+}
+
+impl<T> RoundedRectangle<T>
+where
+    T: num_traits::AsPrimitive<f32>
+        + std::cmp::PartialOrd
+        + std::ops::Add<Output = T>
+        + std::ops::Sub<Output = T>
+        + std::ops::Mul<Output = T>
+        + std::ops::Neg<Output = T>
+        + std::ops::Div<Output = f32>
+        + std::ops::Div<f32, Output = T>
+{
+    /// Returns true if the specified point is inside this rounded rectangle.
+    /// Note: this is always inclusive, in contrast to the `contains` method
+    /// of `Rect` which is sometimes exclusive.
+    #[inline]
+    #[must_use]
+    pub fn contains(&self, point: Vector2<T>) -> bool
+    {
+        //if outside the enclosing rectangle then call it a win and don't proceed
+        if point.x <= self.left()
+            || point.x >= self.right()
+            || point.y >= self.bottom()
+            || point.y <= self.top()
+        {
+            return false;
+        }
+
+        //...by looking at the rounded rectangle as 2 rectangles in a cross and 4
+        //...by circles (overlapping rectangles should be slightly better
+        // than 3 rectangles in this case (I think)):
+        //first rectangle:
+        if point.x >= self.top_left.x
+            && point.y >= self.top_left.y + self.radius
+            && point.x <= self.bottom_right.x
+            && point.y <= self.bottom_right.y - self.radius
+        {
+            return true;
+        }
+        //second rectangle:
+        if point.x >= self.top_left.x + self.radius
+            && point.y >= self.top_left.y
+            && point.x <= self.bottom_right.x - self.radius
+            && point.y <= self.bottom_right.y
+        {
+            return true;
+        }
+
+        //check if the point is inside the 4 circles on the corners by getting the
+        // center of the circles and checking if the distance between the point
+        // and the center is smaller than the radius
+        if (self.top_left + Vector2::new(self.radius, self.radius) - point).magnitude()
```

Comment:
The `sqrt` in `magnitude` is expensive -- a trick I've used in various places is to square the radius instead :) So rather than comparing:

```
sqrt(x*x + y*y) < r
```

We can compare:

```
x*x + y*y < r*r
```

Reply:
Oh damn, that also fixes a trait bound hell. I love this.